### PR TITLE
Refactor Dockerfiles to Use Config Files via ARG

### PR DESCRIPTION
**Pull request by atheo89**
Related to: https://issues.redhat.com/browse/RHAIENG-779

## Description

This PR updates all Dockerfiles to consume configuration files via ARG. Currently, these configuration files define the BASE_IMAGE that each build is based on.

**Changes included:**

- Updated all Dockerfiles to use ARGs. Example: [pytorch+llmcompressor commit](https://github.com/opendatahub-io/notebooks/commit/6027bcfac9330c520fcec3e9636bc36f1e26bdd0). (All related changes are grouped into a single commit for easier review and tracking.)

- Removed CUDA and ROCm layers from Dockerfiles, as these are now provided through the base-images.

- Updated each pipeline to use the `build-args-file` option and consume the corresponding `*.conf` during build.

- Removed the cuda/ folder (now located under [base-images](https://github.com/opendatahub-io/notebooks/tree/main/base-images/cuda/12.6/cuda-repos)).

- Updated the Makefile `build_image` function to consume *.conf files during build.

- Updated `generate_pull_request_pipelineruns.py` to include the `build-args-file` option in the *-pull-request.yaml.

NOTE: The RStudio rhel flavor is not included on this update 

## How Has This Been Tested?

Locally and GHA

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

## Summary by CodeRabbit

- New Features
  - Per-image build-args files and BASE_IMAGE support for container builds.
  - Pipelines now accept a build-args-file parameter.

- Improvements
  - Build system auto-loads build args and passes them to image builds.
  - Pipeline triggers expanded to rerun when build-args files change.
  - CUDA/ROCm images simplified to rely on prebuilt base images.
  - Added labels to aid PipelineRun selection.

- Removals
  - Embedded CUDA/ROCm setup and NVIDIA license/repo entries removed.

### CODE REVIEW

File: .tekton/odh-pipeline-runtime-datascience-cpu-py312-ubi9-pull-request.yaml [hunk 1]
```diff
@@ -13,11 +13,7 @@ metadata:
     pipelinesascode.tekton.dev/max-keep-runs: '3'
     pipelinesascode.tekton.dev/on-comment: ^/kfbuild\s+(all|odh\-pipeline\-runtime\-datascience\-cpu\-py312\-ubi9|runtimes/datascience/ubi9\-python\-3\.12)
     pipelinesascode.tekton.dev/on-cel-expression: |
-      event == "pull_request" && target_branch
-      == "main"
-      && !("manifests/base/params-latest.env".pathChanged())
-      && ( "runtimes/datascience/ubi9-python-3.12/**".pathChanged() ||
-           ".tekton/odh-pipeline-runtime-datascience-cpu-py312-ubi9-pull-request.yaml".pathChanged() )
+      event == "pull_request" && target_branch == "main" && !("manifests/base/params-latest.env".pathChanged()) && ( "runtimes/datascience/ubi9-python-3.12/**".pathChanged() || "runtimes/datascience/ubi9-python-3.12/build-args/cpu.conf".pathChanged() || ".tekton/odh-pipeline-runtime-datascience-cpu-py312-ubi9-pull-request.yaml".pathChanged() )
```
[the remark this answers is not in the document]
Comment (atheo89) [reply .tekton/odh-pipeline-runtime-datascience-cpu-py312-ubi9-pull-request.yaml:16]: Yeah, it changes because gets the format from the `*-push.yaml`

File: .tekton/odh-pipeline-runtime-tensorflow-rocm-py312-ubi9-push.yaml [hunk 2]
```diff
@@ -9,7 +9,7 @@ metadata:
     pipelinesascode.tekton.dev/cancel-in-progress: "false"
     pipelinesascode.tekton.dev/max-keep-runs: "3"
     build.appstudio.openshift.io/build-nudge-files: "manifests/base/params-latest.env"
-    pipelinesascode.tekton.dev/on-cel-expression: event == "push" && target_branch == "main" && !("manifests/base/params-latest.env".pathChanged()) && ( ".tekton/odh-pipeline-runtime-tensorflow-rocm-py312-ubi9-push.yaml".pathChanged() || "runtimes/rocm-tensorflow/ubi9-python-3.12/**".pathChanged() )
+    pipelinesascode.tekton.dev/on-cel-expression: event == "push" && target_branch == "main" && !("manifests/base/params-latest.env".pathChanged()) && ( ".tekton/odh-pipeline-runtime-tensorflow-rocm-py312-ubi9-push.yaml".pathChanged() || "runtimes/rocm-tensorflow/ubi9-python-3.12/**".pathChanged() || "runtimes/rocm-tensorflow/ubi9-python-3.12/build-args/rocm.conf".pathChanged() )
```
Comment (jiridanek) [inline .tekton/odh-pipeline-runtime-tensorflow-rocm-py312-ubi9-push.yaml:12]: every image already has something like

    "runtimes/rocm-tensorflow/ubi9-python-3.12/**".pathChanged()

so maybe

    "runtimes/rocm-tensorflow/ubi9-python-3.12/build-args/rocm.conf".pathChanged()

need not be added explicitly as it is contained in the above?

File: jupyter/minimal/ubi9-python-3.12/Dockerfile.cpu [hunk 3]
```diff
@@ -1,7 +1,12 @@
...
+# configuration args    #
+#########################
+ARG BASE_IMAGE
+
 ####################
-# base             #
+# cpu-base         #
 ####################
-FROM registry.access.redhat.com/ubi9/python-312:latest AS base
+FROM ${BASE_IMAGE} AS cpu-base
```
Comment (jiridanek) [inline jupyter/minimal/ubi9-python-3.12/Dockerfile.cpu:9]: I'm thinking if it would not make sense to define

* CPU_BASE_IMAGE
* CUDA_BASE_IMAGE
* ROCM_BASE_IMAGE

so that they can be all defined in a single .conf file alltogether

may be a bit easier to maintain later

but no strong preference either way

Comment (atheo89) [reply jupyter/minimal/ubi9-python-3.12/Dockerfile.cpu:9]: Yes, I thought the same, and it makes sense. However, I moved to using a single .conf file for each image so that, if we decide to add specific parameters in the future, we’ll have more flexibility.

File: scripts/generate_pull_request_pipelineruns.py
Comment (jiridanek) [inline scripts/generate_pull_request_pipelineruns.py:1]: @coderabbitai explain to me the changes made in this file

File: .tekton/odh-pipeline-runtime-datascience-cpu-py312-ubi9-pull-request.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jiridanek) [reply .tekton/odh-pipeline-runtime-datascience-cpu-py312-ubi9-pull-request.yaml:16]: I see

File: Makefile [hunk 4]
```diff
@@ -67,9 +67,20 @@ endif
...
 	$(eval IMAGE_NAME := $(IMAGE_REGISTRY):$(1)-$(IMAGE_TAG))
-	$(eval BUILD_ARGS :=)
 
-	$(info # Building $(IMAGE_NAME) image...)
+	# Checks if there’s a build-args/*.conf matching the Dockerfile
+	$(eval BUILD_DIR := $(dir $(2)))
+	$(eval DOCKERFILE_NAME := $(notdir $(2)))
+	$(eval CONF_FILE := $(BUILD_DIR)build-args/$(shell echo $(DOCKERFILE_NAME) | cut -d. -f2).conf)
+
+	# if the conf file exists, transform it into --build-arg KEY=VALUE flags
```
[the remark this answers is not in the document]
Comment (atheo89) [reply Makefile:76]: Yes, i utilized it here: 	`$(CONTAINER_ENGINE) build $(CONTAINER_BUILD_CACHE_ARGS) --platform=$(BUILD_ARCH) --label release=$(RELEASE) --tag $(IMAGE_NAME) --file '$(2)' $(BUILD_ARGS) {}\;`

and the corresponding extraction happen here: 	
```
# if the conf file exists, transform it into --build-arg KEY=VALUE flags
	$(eval BUILD_ARGS := $(shell if [ -f $(CONF_FILE) ]; then \
		while IFS='=' read -r k v; do \
			[ -n "$$k" ] && printf -- "--build-arg %s=%s " "$$k" "$$v"; \
		done < $(CONF_FILE); \
	fi))
```

File: .tekton/odh-workbench-rstudio-minimal-cpu-py311-c9s-push.yaml [hunk 5]
```diff
@@ -29,6 +29,8 @@ spec:
     value: quay.io/opendatahub/odh-workbench-rstudio-minimal-cpu-py311-c9s:{{revision}}
   - name: dockerfile
     value: rstudio/c9s-python-3.11/Dockerfile.cpu
+  - name: build-args-file
+    value: rstudio/c9s-python-3.11/build-args/cpu.conf
```
[the remark this answers is not in the document]
Comment (atheo89) [reply .tekton/odh-workbench-rstudio-minimal-cpu-py311-c9s-push.yaml:33]: it is already declared on singlearch-push-pipeline.yaml file

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (jiridanek) [reply Makefile:76]: u're using --build-arg, not --build-arg-file, but ok, works too
